### PR TITLE
[SYCL] Add FPGA specific command line interfaces for AOT

### PR DESCRIPTION
**Pull request by mdtoguchi**
Adds specific interfaces which is used by the FPGA folks to enable AOT
compilation behaviors.  This mainly is a simpler interface over the existing
-fsycl -fsycl-targets enabling.

-fintelfpga - enables FPGA AOT compilation flow
-Xs <arg> - short way of passing options to backend compiler
-fsycl-link=early/image - enable creation of aocx/aocr fat binaries

Signed-off-by: Michael D Toguchi <michael.d.toguchi@intel.com>

### CODE REVIEW

File: clang/include/clang/Driver/Action.h [hunk 1]
```diff
@@ -639,6 +640,17 @@ class SPIRVTranslatorJobAction : public JobAction {
   }
 };
 
+class SPIRCheckJobAction : public JobAction {
```
Comment (kbobrovs) [inline clang/include/clang/Driver/Action.h]: please add a comment what this job is for

File: clang/include/clang/Driver/Types.def [hunk 2]
```diff
@@ -106,4 +106,6 @@ TYPE("sycl-header",              SYCL_Header,  INVALID,         "h",     "")
 TYPE("sycl-fatbin",              SYCL_FATBIN,  INVALID,         nullptr, "")
 TYPE("hip-fatbin",               HIP_FATBIN,   INVALID,         "hipfb", "A")
 TYPE("tempfilelist",             Tempfilelist, INVALID,         nullptr, "")
+TYPE("aocx",                     AOCX,         INVALID,         "aocx",  "")
```
Comment (kbobrovs) [inline clang/include/clang/Driver/Types.def]: I suggest to add a clarifying prefix, like fpga-aocx

File: clang/include/clang/Driver/Action.h (hunk 1 above)
Comment (kbobrovs) [reply clang/include/clang/Driver/Action.h]: Currently it is used only for FPGA-specific checking, but I think this action/tool is the place to add more checks of a SPIRV code in future if needed. - could be added to the comment too.

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -6365,6 +6365,28 @@ void OffloadBundler::ConstructJob(Compilation &C, const JobAction &JA,
...
+  // and host object inputs
+  const ToolChain *TCCheck = &getToolChain();
+  if (TCArgs.hasArg(options::OPT_fsycl_link_EQ) &&
+      TCCheck->getTriple().getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
+    Triples = "-targets=";
+    llvm::Triple TT;
+    TT.setArchName(JA.getInputs()[0]->getType() == types::TY_AOCX ? "fpga_aocx"
+                                                                 : "fpga_aocr");
+    TT.setVendorName("intel");
+    TT.setOS(llvm::Triple(llvm::sys::getProcessTriple()).getOS());
```
Comment (kbobrovs) [inline clang/lib/Driver/ToolChains/Clang.cpp]: AFAIU, we can do a cross compilation, so should not the OS component from the from the TCCheck->getTriple() be used?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 4]
```diff
@@ -6421,8 +6443,8 @@ void OffloadBundler::ConstructJobMultipleOutputs(
   // contain bundled objects). We will perform partial linking against the
   // object and specific offload target archives which will be sent to the
   // unbundler to produce a list of target objects.
-  if (Input.getType() == types::TY_Object &&
-      TCArgs.hasArg(options::OPT_foffload_static_lib_EQ)) {
+  bool StaticLib = TCArgs.hasArg(options::OPT_foffload_static_lib_EQ);
+  if (Input.getType() == types::TY_Object && StaticLib) {
```
Comment (kbobrovs) [inline clang/lib/Driver/ToolChains/Clang.cpp]: StaticLib seems to be still used just once - unnecessary change?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 5]
```diff
@@ -6456,6 +6482,22 @@ void OffloadBundler::ConstructJobMultipleOutputs(
...
+    // aocx or aocr type bundles.
+    if (Dep.DependentToolChain->getTriple().getSubArch() ==
+                                   llvm::Triple::SPIRSubArch_fpga &&
+        (Input.getType() == types::TY_AOCX ||
+         Input.getType() == types::TY_AOCR)) {
+      llvm::Triple TT;
+      TT.setArchName(Input.getType() == types::TY_AOCX ? "fpga_aocx"
+                                                       : "fpga_aocr");
+      TT.setVendorName("intel");
+      TT.setOS(llvm::Triple(llvm::sys::getProcessTriple()).getOS());
```
Comment (kbobrovs) [inline clang/lib/Driver/ToolChains/Clang.cpp]: please see comment above re OS component

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 6]
```diff
@@ -6587,3 +6629,26 @@ void SPIRVTranslator::ConstructJob(Compilation &C, const JobAction &JA,
...
+                             const InputInfo &Output,
+                             const InputInfoList &Inputs,
+                             const llvm::opt::ArgList &TCArgs,
+                             const char *LinkingOutput) const {
+  // Construct llvm-no-spir-kernel command.
+  assert(isa<SPIRCheckJobAction>(JA) && "Expecting SPIR Check job!");
+
+  // The translator command looks like this:
+  // llvm-no-spir-kernel <file>.bc
+  // Upon success, we just move ahead.  Error means a kernel was found and
```
Comment (kbobrovs) [inline clang/lib/Driver/ToolChains/Clang.cpp]: > Error means a kernel was found and

I suggest to treat the checker as the black-box here, as it may evolve - "Error means the check failed"

File: clang/lib/Driver/ToolChains/Gnu.cpp [hunk 7]
```diff
@@ -587,8 +587,13 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
 
       AddRunTimeLibs(ToolChain, D, CmdArgs, Args);
 
-      if (Args.hasArg(options::OPT_fsycl))
+      if (Args.hasArg(options::OPT_fsycl)) {
         CmdArgs.push_back("-lsycl");
+        // Use of -fintelfpga implies -lOpenCL.
+        // FIXME: Adjust to use plugin interface when available.
+        if (Args.hasArg(options::OPT_fintelfpga))
+          CmdArgs.push_back("-lOpenCL");
```
Comment (kbobrovs) [inline clang/lib/Driver/ToolChains/Gnu.cpp]: Useful change! Can we please make it default for all targets for now (until plugin interface is in place)?

File: clang/lib/Driver/Driver.cpp [hunk 8]
```diff
@@ -3223,6 +3306,35 @@ class OffloadingActionBuilder final {
 
         unsigned I = 0;
         for (auto &LI : DeviceLinkerInputs) {
+          // Perform a check for device kernels.  This is done for FPGA with
```
Comment (agozillon) [inline clang/lib/Driver/Driver.cpp]: can possibly remove the word "with"

File: clang/lib/Driver/Driver.cpp [hunk 9]
```diff
@@ -3322,6 +3438,16 @@ class OffloadingActionBuilder final {
...
                                               options::OPT_fno_sycl, false);
+
+      // Add additional check for -fintelfpga, which if set will add the fpga
+      // specific target triple.  There is a check done earlier for the setting
+      // of both options, so we can freely just check argument contents.
+      if (!SYCLTargets && C.getInputArgs().hasArg(options::OPT_fintelfpga)) {
+        const char * AStr =
+                Args.MakeArgString("spir64_fpga-unknown-linux-sycldevice");
+        SYCLTargets = Args.MakeJoinedArg(nullptr,
+        C.getDriver().getOpts().getOption(options::OPT_fsycl_targets_EQ), AStr);
```
Comment (agozillon) [inline clang/lib/Driver/Driver.cpp]: Indentation seems a little off here

File: clang/lib/Driver/Driver.cpp [hunk 10]
```diff
@@ -4720,6 +4914,18 @@ InputInfo Driver::BuildJobsForActionNoCache(
                         C.addTempFile(C.getArgs().MakeArgString(TmpFileName),
                                       types::TY_Tempfilelist);
         CurI = InputInfo(types::TY_Tempfilelist, TmpFile, TmpFile);
+      } else if (EffectiveTriple.getSubArch() ==
+                                         llvm::Triple::SPIRSubArch_fpga &&
```
Comment (agozillon) [inline clang/lib/Driver/Driver.cpp]: Indentation is slightly off I think

File: clang/lib/Driver/Driver.cpp (hunk 8 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp]: yeah, 'with when' doesn't work.

File: clang/lib/Driver/Driver.cpp (hunk 9 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp]: I tried to keep formatting within 80 columns which is why it is at the same level.  I'll do some slight refactoring.

File: clang/lib/Driver/Driver.cpp (hunk 10 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp]: Another 'let me fit this in 80 columns' format weirdness.  I'll change the ordering so it fits better.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 6 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: Will adjust.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 4 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: It was necessary, until I made some changes.  I'll clean out.

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 3 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: makes since - will adjust.

**Comment by mdtoguchi** [comment]
rebased to fix conflicts from other merges.

File: clang/lib/Driver/Driver.cpp [hunk 11]
```diff
@@ -801,14 +816,19 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
...
     if (HasValidSYCLRuntime) {
       const ToolChain *HostTC =
           C.getSingleOffloadToolChain<Action::OFK_Host>();
       llvm::Triple TT(TargetTriple);
-      TT.setArch(llvm::Triple::spir64);
-      TT.setVendor(llvm::Triple::UnknownVendor);
-      TT.setOS(llvm::Triple(llvm::sys::getProcessTriple()).getOS());
-      TT.setEnvironment(llvm::Triple::SYCLDevice);
+      if (SYCLfpga)
+        TT.setTriple("spir64_fpga-unknown-linux-sycldevice");
```
Comment (kbobrovs) [inline clang/lib/Driver/Driver.cpp]: this probably won't work for Windows. I was thinking about using the triple components provided in -fsycl-targets, including OS, rather then overriding it here. If no OS was provided, then llvm::Triple(llvm::sys::getProcessTriple()).getOS()) seems like the proper choice. Does this sound reasonable?

Comment (bader) [reply clang/lib/Driver/Driver.cpp]: The OS component of target triple should be unknown, as devices do not have any OS (as well as environment/runtime).
This information should be inferred from the aux-triple, which defines the host OS and environment.

I suggest fixing this in a separate patch, as it might require changes in other components also, not just the driver.

Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp]: I will update the default triple for -fintelfpga to match what Konst has suggested.  We can make adjustments using 'unknown' for the OS later as Alexey stated.
